### ISSUE TITLE
runtime error when using normalization layer in network

### ISSUE THREAD

**Issue by uu-praveeng** [ARM-software/armnn#307]
Hi ,

I am trying to create a mnist network in which there is a sequence of FCLayer followed by cross channel normalization .

Attached file([mnist.cpp.tar.gz](https://github.com/ARM-software/armnn/files/3911352/mnist.cpp.tar.gz)) shows the network creation workflow that I have used. When I have executed the exe on hikey960 using armnn1908 libs I am seeing the below error

```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at

Thread 1 "mnist_exe" received signal SIGABRT, Aborted.
0x0000007fbe492528 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt

#0  0x0000007fbe492528 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x0000007fbe4939e0 in __GI_abort () at abort.c:89
#2  0x0000007fbe66c254 in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#3  0x0000007fbe669dc4 in ?? () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#4  0x0000007fbe669e10 in std::terminate() () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#5  0x0000007fbe66a0d4 in __cxa_throw () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#6  0x0000007fbe665b98 in std::__throw_out_of_range(char const*) () from /usr/lib/aarch64-linux-gnu/libstdc++.so.6
#7  0x0000007fbeab17dc in armnn::Runtime::GetLoadedNetworkPtr(int) const () from /home/arm01/armnn-devenv/Praveen/armnn1908/build/libarmnn.so
#8  0x0000007fbeab1840 in armnn::Runtime::GetInputTensorInfo(int, int) const () from /home/arm01/armnn-devenv/Praveen/armnn1908/build/libarmnn.so
#9  0x0000000000403de8 in main () at mnist.cpp:291
```

Let me know if I have configured any layer wrongly. Any help would be highly appreciated.

#### Note:Layer outputs are as expected until ReLULayer. 

Thank you,
Praveen.

**Comment by MikeJKelly** [comment]
Hi Praveen,

What kind of device are you running this on? I get an exception like the one you're seeing if I try to run CpuAcc on my x86 chip, during the Optimize(...) method no supplied backend is available so the method returns a null. 

If I switch CpuAcc to CpuRef it gets past that issue and I see an error where the Normalization layer seems to have a two dimensional input shape {1, 500} while the Armnn workload expects a 4 dimensional one (it expects an NCHW or NHWC layout). 

I hope that helps?

Best regards,
Mike

**Comment by uu-praveeng** [comment]
Hi Mike,

I used HiKey960.

- with CpuAcc I have seen the above mentioned error 

- with CpuRef I have seen the num dimensions issue. 

Is it mandated to that normalization should take 4 dimensional input?
If yes, do I need to use the reshape layer before Norm Layer?

I did try to create the standalone setup with input and normalization layer as a standalone setup with 4 dimension input. However output of normalization layer is not same as one got using the other libraries like mkldnn,cudnn. 
Could you assist me in this, I am unclear what I am missing here.
Attached is the standalone setup for your reference [link](https://github.com/ARM-software/armnn/files/3921939/norm.zip)

Thank you,
Praveen.

**Comment by MikeJKelly** [comment]
Hi Praveen, 

Armnn requires that the Normalization layer have a 4 dimensional input. I wasn't sure if that would work aso I added a reshape layer (Reshaping the output of the layer to {1,1,1,500}, which I think should be OK) between the ReLU layer and the Normalization layer and the graph worked for me (outputting 500 floats). I've attached my version of 
[mnist.cpp.gz](https://github.com/ARM-software/armnn/files/3922949/mnist.cpp.gz)
.

**Comment by uu-praveeng** [comment]
Hi Mike,

Thanks for looking into this.
In the previous **link**  I have shared . It as the standalone normlayer along with input(mw_interm_tap6.txt) and expected output(mw_inerm_tap7.txt) . 

I observed  different results than the expected output. Could you cross check that as well.
Could it be ACL impl issue? 

Thank you,
Praveen.

**Comment by TeresaARM** [comment]
Hi Praveen,

Thank you for your answer.

Running a Google Colab notebook the following code:
```
import tensorflow as tf
import numpy as np
sess = tf.Session()
data = np.loadtxt('mw_interm_tap_6.txt', dtype=str)
input=tf.constant(data.astype(float).tolist())
input = tf.reshape(input, [1,1,1,500]) 
radius = 2
bias = 1.0
alpha = 0.0001
beta = 0.75
tf_opreation0 = tf.nn.local_response_normalization(
    input,
    depth_radius=radius,
    bias=bias,
    alpha=alpha,
    beta=beta,
    name=None
)
tf_result0 = sess.run(tf_opreation0)
np.set_printoptions(suppress=True)
print (np.round(tf_result0, decimals=5))
```
same values as in _NormStandalone.txt_ are obtained. 

How did you obtain the values in _mw_interm_tap_7.txt_ ? 
Which _TensorFlow_ or _PyTorch_ function would give the values in _mw_interm_tap_7.txt_?

Reference:
https://www.tensorflow.org/api_docs/python/tf/nn/local_response_normalization

**Comment by uu-praveeng** [comment]
Hi Teresa,

I did not verify with tensorflow or pyTorch function calls. 
But if we use the ACL standalone function call NENormalizationLayer with same configuration params and inputs, I would get the output same as mw_interm_tap_7.txt.

Let me know if you need the standalone setup for ACL Normalization Layer? 
I am can share  that for you.

Thank you,
Praveen.

**Comment by MikeJKelly** [comment]
Hi Praveen, 

I was looking at this yesterday and I'm getting the same results as you, @TeresaARM also tried your test scenario on Tensorflow and got the same results, the Android NNAPI tests also all pass. So I believe that the ACL implementation is working correctly. 

I'm not familiar with mkldnn or cudnn could they have a different implementation of Local Response Normalization or use its parameters in a different way?

Best regards,
Mike

**Comment by TeresaARM** [comment]
Hi Praveen,
could you send us the ACL standalone function? are you using exactly same values for both armnn standalone and ACL standalone? Input parameters might be different (radius vs. size of the kernel("diameter"))
Regards

**Comment by uu-praveeng** [comment]
Hi @TeresaARM ,

Attached the standalone test setup for NeNormalization layer using ACL.
Let me know if you need any further info.

[NeNormlization.zip](https://github.com/ARM-software/armnn/files/3944506/NeNormlization.zip)

Thank you,
Praveen.

**Comment by TeresaARM** [comment]
Hi Praveen,
So far I could see that:
in Tensorflow I get Armnn results when I put this input:
- radius = 2
- bias = 1.0
- alpha = 0.0001
- beta = 0.75
and ACL result when I put these ones:
- radius = 2
- norm_size = radius*2 +1
- bias = 1.0
- alpha = 0.0001 / norm_size
- beta = 0.75
Thank you

**Comment by uu-praveeng** [comment]
Hi @TeresaARM 

I assume ACL and ARMNN outputs should be same since the ACL is backend for ARMNN.
If no, is there anyway we could make ACL and ARMNN outputs same with configuration changes in either of the one?

Thank youi,
Praveen.

**Comment by TeresaARM** [comment]
Hi Praveen,
both implementations are right. In the class _NormalizationLayerInfo_ there is a parameter with its correspondent comment:
_@param[in] is_scaled (Optional) Boolean that specifies if alpha will be scaled by the normalization size or not. Should be false to follow [Krichevksy 2012]._
This boolean allows you to choose which of both implementations you want to execute in ACL. In armnn, in the call to the ACL normalization layer, this parameter is set to 'false', in your code is set to 'true'. In the armnn reference implementation there is only one option.
Please let us know if this information covers your question.

**Comment by uu-praveeng** [comment]
Hi @TeresaARM ,

Are there any plans to add this argument to NormLayer API in armnn.

Thank you,
Praveen.

**Comment by matthewsloyanARM** [comment]
Hi @uu-praveeng,

Thank you for reaching out! Based on the updated `README.md` found [here](https://github.com/ARM-software/armnn/tree/master#how-to-get-involved) under the "How to get involved" section we will be closing GitHub issues labeled with "Feature Request". Instead, we will be tracking the issue and adding the "Help wanted" label. All the issues with this label can be found [here](https://github.com/ARM-software/armnn/issues?q=is%3Aissue+label%3A%22Help+wanted%22+). We would much appreciate your help to add these features. If you would like to add this or any feature to Arm NN feel free to reopen the ticket and let us know you are working on it with a quick comment. We will then track the updates and help you get your additions into our codebase.

Below you will find our contributors guide which will show you how you can submit a patch to us. If you have any questions feel free to ask here or by creating a new Issue. Thank you!

https://www.mlplatform.org/contributing

Kind regards,

Matthew
